tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,367 @@
+        rate_matrix_block,
+        atomic_number,
+        ion_number,
+        total_rad_recomb_coefficients,
+        total_photo_ion_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """Calculates the row for the species treated in NLTE ionization
+
+        Parameters
+        ----------
+        rate_matrix_block : numpy.array
+            The diagonal block corresponding to current atomic number.
+        atomic_number : int
+            Current atomic number
+        ion_number : int
+            Current ion number
+        total_rad_recomb_coefficients : DataFrame
+            Rad. recomb. coefficients for current atomic number
+        total_photo_ion_coefficients : DataFrame
+            Photo ion. coefficients for current atomic number
+        total_coll_ion_coefficients : DataFrame
+            Coll. ion. coefficients for current atomic number
+        total_coll_recomb_coefficients : DataFrame
+            Coll. recomb. coefficients for current atomic number
+
+        Returns
+        -------
+        numpy.array
+            Rate matrix block with a changed row for NLTE ionization treatment
+        """
+        ion_coefficients = (
+            total_photo_ion_coefficients + total_coll_ion_coefficients
+        )
+        recomb_coefficients = (
+            total_rad_recomb_coefficients + total_coll_recomb_coefficients
+        )
+        if atomic_number != ion_number:
+            ion_coeff_matrix_ion_row = NLTERateEquationSolver.ion_matrix(
+                ion_coefficients, atomic_number, ion_number
+            )
+            recomb_coeff_matrix_ion_row = NLTERateEquationSolver.recomb_matrix(
+                recomb_coefficients, atomic_number, ion_number
+            )
+            rate_matrix_block[ion_number, :] = (
+                ion_coeff_matrix_ion_row + recomb_coeff_matrix_ion_row
+            )
+        return rate_matrix_block
+
+    @staticmethod
+    def lte_rate_matrix_block(phi_block, electron_density):
+        """Creates the generic LTE block for rate matrix.
+
+        Parameters
+        ----------
+        phi_block : DataFrame
+            Saha Factors for current atomic number
+        electron_density : float
+            Current guess for electron density
+
+        Returns
+        -------
+        numpy.array
+            LTE block for rate matrix
+        """
+        lte_rate_vector_block = -1.0 * np.hstack([*phi_block.values, -1.0])
+        lte_rate_matrix_block = np.diag(lte_rate_vector_block)
+        n_e_initial = np.ones(len(phi_block)) * electron_density
+        n_e_matrix = np.diag(n_e_initial, 1)
+        lte_rate_matrix_block += n_e_matrix
+        lte_rate_matrix_block[-1, :] = 1.0
+        return lte_rate_matrix_block
+
+    @staticmethod
+    def prepare_phi(phi):
+        """
+        Makes sure that phi does not have any 0 entries.
+        """
+        phi[phi == 0.0] = 1.0e-10 * phi[phi > 0.0].min().min()
```

Comment:
Can we not just use a constant small number instead of scaling the minimum value in the table?

Reply:
This is done to make sure that "zero" elements have significantly less "weight" than the other Saha factors. Making it constant might be an issue if Saha factors are generally small and the constant we choose might be greater than some of them.